internal/device/geopm_cpu_power_meter.go

```diff
@@ -0,0 +1,127 @@
+// SPDX-FileCopyrightText: 2025 The Kepler Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package device
+
+import (
+	"fmt"
+	"log/slog"
+	"sync"
+	"github.com/geopm/geopm/geopmdgo/geopmdgo"
+)
+
+// geopmEnergyZone represents a single energy zone using the GEOPM PIO interface.
+type geopmEnergyZone struct {
+	name      string
+	index     int
+	domain    int
+	mu        sync.Mutex
+}
+
+// Name returns the name of the energy zone.
+func (z *geopmEnergyZone) Name() string {
+	return z.name
+}
+
+// Index returns the index of the energy zone.
+func (z *geopmEnergyZone) Index() int {
+	return z.index
+}
+
+// Path returns an empty string as GEOPM does not use sysfs paths.
+func (z *geopmEnergyZone) Path() string {
+	return ""
+}
+
+// Energy reads the energy consumed by the zone using the GEOPM PIO interface.
+func (z *geopmEnergyZone) Energy() (Energy, error) {
+	z.mu.Lock()
+	defer z.mu.Unlock()
+
+	value, err := geopmdgo.ReadSignal("CPU_ENERGY", z.domain, z.index)
+	if err != nil {
+		return 0, fmt.Errorf("failed to read energy for zone %s: %w", z.name, err)
+	}
+	return Energy(value), nil
+}
+
+// MaxEnergy returns 0 as GEOPM does not provide a maximum energy value.
+func (z *geopmEnergyZone) MaxEnergy() Energy {
+	return 0
```

Comment:
Does geopm keep incrementing until overflow? In that case, can we keep it as MaxUint64?

Reply:
I will adjust the implementation to use the PushSignal()/Sample() API.  This will handle the rollover internally to GEOPM, and also provide a faster read mechanism when io-uring is enabled. Note the return type of Sample() is a double.

Reply:
Sorry for the confusion.  I didn't really get a chance to think this through today during the churn of meetings.  The rollover is handled by GEOPM for both the ReadSignal and PushSignal/Sample() APIs as long as the calls are within the same process.

Reply:
I've had a very busy day today, and should not have responded to your question without some more thought.

1. Thanks for the feedback, it is helped me create this bug:
https://github.com/geopm/geopm/issues/3914

2. The intention is that GEOPM will manage the roll over for these counters for calls within a single process, either using the ReadSignal() or PushSignal()/Sample() APIs

3. The rollover is handled if the support is provided by the msr device driver interface, however this is not the default.  By default the powercap device driver interface is used, and the rollover handling was not implemented for the powercap driver.

I will work to implement the fix for the powercap interface.